src/coreclr/vm/pgo.cpp

```diff
@@ -285,206 +322,516 @@ void PgoManager::ReadPgoData()
+    *offsetOfActualInstrumentationData = (UINT)AlignUp(headerSize, maxAlign);
     return S_OK;
 }
 
-HRESULT PgoManager::getMethodBlockCounts(MethodDesc* pMD, unsigned ilSize, UINT32* pCount,
-    ICorJitInfo::BlockCounts** pBlockCounts, UINT32* pNumRuns)
+HRESULT PgoManager::allocPgoInstrumentationBySchemaInstance(MethodDesc* pMD,
+                                                            PgoInstrumentationSchema* pSchema,
+                                                            UINT32 countSchemaItems,
+                                                            BYTE** pInstrumentationData)
 {
-    // Initialize our out params
-    *pCount = 0;
-    *pBlockCounts = NULL;
-    *pNumRuns = 0;
-
-   // Bail if there's no profile data.
-    //
-    if (s_PgoData == NULL)
+    // Initialize our out param
+    *pInstrumentationData = NULL;
+    int codehash;
+    unsigned ilSize;
+    if (!GetVersionResilientILCodeHashCode(pMD, &codehash, &ilSize))
     {
         return E_NOTIMPL;
     }
 
-    // See if we can find counts for this method in the profile buffer.
-    //
-    const unsigned maxIndex = s_PgoIndex;
-    const unsigned token    = pMD->IsDynamicMethod() ? 0 : pMD->GetMemberDef();
-    const unsigned hash     = pMD->GetStableHash();
+    UINT offsetOfActualInstrumentationData;
+    HRESULT hr = ComputeOffsetOfActualInstrumentationData(pSchema, countSchemaItems, sizeof(HeaderList), &offsetOfActualInstrumentationData);
+    UINT offsetOfInstrumentationDataFromStartOfDataRegion = offsetOfActualInstrumentationData - sizeof(HeaderList);
+    if (FAILED(hr))
+    {
+        return hr;
+    }
 
+    // Compute offsets for each instrumentation entry
+    PgoInstrumentationSchema prevSchema;
+    memset(&prevSchema, 0, sizeof(PgoInstrumentationSchema));
+    prevSchema.Offset = offsetOfInstrumentationDataFromStartOfDataRegion;
+    for (UINT32 iSchema = 0; iSchema < countSchemaItems; iSchema++)
+    {
+        LayoutPgoInstrumentationSchema(prevSchema, &pSchema[iSchema]);
+        prevSchema = pSchema[iSchema];
+    }
 
-    unsigned index = 0;
-    unsigned methodsChecked = 0;
+    S_SIZE_T allocationSize = S_SIZE_T(sizeof(HeaderList)) + S_SIZE_T(pSchema[countSchemaItems - 1].Offset + InstrumentationKindToSize(pSchema[countSchemaItems - 1].InstrumentationKind));
+    
+    if (allocationSize.IsOverflow())
+    {
+        return E_NOTIMPL;
+    }
+    size_t unsafeAllocationSize = allocationSize.Value();
+    HeaderList* pHeaderList = NULL;
 
-    while (index < maxIndex)
+    if (pMD->IsDynamicMethod())
     {
-        // The first two "records" of each entry are actually header data
-        // to identify the method.
-        //
-        Header* const header = (Header*)&s_PgoData[index];
+        HeaderList *currentHeaderList = m_pgoHeaders;
+        if (currentHeaderList != NULL)
+        {
+            if (!ComparePgoSchemaEquals(currentHeaderList->header.GetData(), currentHeaderList->header.countsOffset, pSchema, countSchemaItems))
```

Comment:
Does this need to tie into the IL versioning scheme? I don't think the jit knows which IL version is active for a jit request, but presumably the jit interface does (or could know).

Reply:
I wouldn't expect so. The pgo instrumentation data structures will be the same if the schema's match up, and the JIT already must tolerate data which isn't completely accurate.

Reply:
Now, the reading side may be more interesting, but again, that gets into the question of how bad is it to present incorrectly structured data, if that's really bad we could easily add the IL code hash into the schema to allow for the JIT to check it for correctness or something.